File: packages/amplify_auth_cognito/ios/Classes/FlutterConfirmSignInRequest.swift
/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import Foundation
import Amplify
import AmplifyPlugins
import amplify_core

struct FlutterConfirmSignInRequest {
  var confirmationCode: String
  var options: AuthConfirmSignInOperation.Request.Options?

  init(dict: NSMutableDictionary){
    self.confirmationCode = dict["confirmationCode"] as! String
    self.options = formatOptions(options: dict["options"] as! Dictionary<String, Any>?)

  }
    
  func formatOptions(options: Dictionary<String, Any>?) -> AuthConfirmSignInOperation.Request.Options {
    let rawAttributes = options?["userAttributes"] as? [String : String] ?? [:]
    let userAttributes: [AuthUserAttribute] = rawAttributes.map { key, value in
      AuthUserAttribute(createAuthUserAttributeKey(keyName: key), value: value)
    }
    
    let pluginOptions =  AWSAuthConfirmSignInOptions(
      userAttributes: userAttributes,
      metadata: options?["clientMetadata"] as? [String : String]
    )
    return AuthConfirmSignInOperation.Request.Options(pluginOptions: pluginOptions)
  }
    
  static func validate(dict: NSMutableDictionary) throws {
    let validationErrorMessage = "ConfirmSignIn Request malformed."
    if (dict["confirmationCode"] == nil && dict["options"] == nil) {
        throw InvalidRequestError.auth(comment: validationErrorMessage,
                                          suggestion: String(format: ErrorMessages.missingAttribute, "confirmationCode"))
    }
  }
}

File: packages/auth/amplify_auth_cognito_android/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/types/FlutterConfirmSignInRequest.kt
/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

package com.amazonaws.amplify.amplify_auth_cognito.types

import com.amazonaws.amplify.amplify_auth_cognito.utils.createAuthUserAttribute
import com.amazonaws.amplify.amplify_core.exception.ExceptionMessages
import com.amazonaws.amplify.amplify_core.exception.InvalidRequestException
import com.amplifyframework.auth.cognito.options.AWSCognitoAuthConfirmSignInOptions

data class FlutterConfirmSignInRequest(val map: HashMap<String, *>) {
    val confirmationCode: String = map["confirmationCode"] as String
    val options: AWSCognitoAuthConfirmSignInOptions = formatOptions(map["options"] as HashMap<String, *>?)

    private fun formatOptions(rawOptions: HashMap<String, *>?): AWSCognitoAuthConfirmSignInOptions {
        val options = AWSCognitoAuthConfirmSignInOptions.builder()

        if (rawOptions?.get("clientMetadata") != null)
            options.metadata(rawOptions["clientMetadata"] as HashMap<String, String>)

        val rawAttributes = rawOptions?.get("userAttributes") as HashMap<String, String>? ?: emptyMap<String, String>()
        val attributes = rawAttributes.map { (key, value) ->
            createAuthUserAttribute(key, value)
        }

        options.userAttributes(attributes)

        return options.build()
    }

    companion object {
        private const val validationErrorMessage: String = "ConfirmSignIn Request malformed."
        fun validate(req: HashMap<String, *>?) {
            if (req == null || req !is HashMap<String, *>) {
                throw InvalidRequestException(validationErrorMessage, ExceptionMessages.missingAttribute.format("request map"))
            }
        }
    }
}
